File: src/ch2d/bindings/sprite.cpp
// ch2d
#include <ch2d/System.hpp>

namespace ch2d
{
    unsigned int System::sprite_create(void)
    {
        return mSpriteHandler.create();
    }

    bool System::sprite_draw(unsigned int id)
    {
        auto sprite = mSpriteHandler.get(id);

        mRenderWindow.draw(*sprite);

        return true;
    }

    bool System::sprite_remove(unsigned int id)
    {
        return mSpriteHandler.remove(id);
    }

    bool System::sprite_setPosition(unsigned int id, LUA_NUMBER x, LUA_NUMBER y)
    {
        auto sprite = mSpriteHandler.get(id);

        if(nullptr == sprite)
        {
            return false;
        }

        sprite->setPosition(sf::Vector2f(std::floor(x), std::floor(y)));

        return true;
    }

    bool System::sprite_setOrigin(unsigned int id, LUA_NUMBER x, LUA_NUMBER y)
    {
        auto sprite = mSpriteHandler.get(id);

        if(nullptr == sprite)
        {
            return false;
        }

        sprite->setOrigin(std::floor(x), std::floor(y));

        return true;
    }

    bool System::sprite_setRotation(unsigned int id, LUA_NUMBER rotation)
    {
        auto sprite = mSpriteHandler.get(id);

        if(nullptr == sprite)
        {
            return false;
        }

        sprite->setRotation(rotation);

        return true;
    }

    bool System::sprite_setScale(unsigned int id, LUA_NUMBER x, LUA_NUMBER y)
    {
        auto sprite = mSpriteHandler.get(id);

        if(nullptr == sprite)
        {
            return false;
        }

        sprite->setScale(x, y);

        return true;
    }

    bool System::sprite_setTexture(unsigned int spriteId, unsigned int textureId)
    {
        auto sprite = mSpriteHandler.get(spriteId);

        auto texture = mTextureHandler.get(textureId);

        if(nullptr == sprite || nullptr == texture)
        {
            return false;
        }

        sprite->setTexture(*texture);

        return true;
    }

    bool System::sprite_setTextureRect(unsigned int id, LUA_NUMBER top, LUA_NUMBER left, LUA_NUMBER width, LUA_NUMBER height)
    {
        auto sprite = mSpriteHandler.get(id);

        if(nullptr == sprite)
        {
            return false;
        }

        sprite->setTextureRect(sf::IntRect(top, left, width, height));

        return true;
    }

    std::tuple<LUA_NUMBER, LUA_NUMBER> System::sprite_getPosition(unsigned int id)
    {
        std::tuple<LUA_NUMBER, LUA_NUMBER> data {0.f, 0.f};

        auto sprite = mSpriteHandler.get(id);

        if(nullptr == sprite)
        {
            return data;
        }

        auto position = sprite->getPosition();

        data = std::make_tuple(position.x, position.y);

        return data;
    }

    std::tuple<LUA_NUMBER, LUA_NUMBER> System::sprite_getOrigin(unsigned int id)
    {
        std::tuple<LUA_NUMBER, LUA_NUMBER> data {0.f, 0.f};

        auto sprite = mSpriteHandler.get(id);

        if(nullptr == sprite)
        {
            return data;
        }

        auto origin = sprite->getOrigin();

        data = std::make_tuple(origin.x, origin.y);

        return data;
    }

    LUA_NUMBER System::sprite_getRotation(unsigned int id)
    {
        LUA_NUMBER data {0.f};

        auto sprite = mSpriteHandler.get(id);

        if(nullptr == sprite)
        {
            return data;
        }

        auto rotation = sprite->getRotation();

        data = rotation;

        return data;
    }

    std::tuple<LUA_NUMBER, LUA_NUMBER> System::sprite_getScale(unsigned int id)
    {
        std::tuple<LUA_NUMBER, LUA_NUMBER> data {0.f, 0.f};

        auto sprite = mSpriteHandler.get(id);

        if(nullptr == sprite)
        {
            return data;
        }

        auto scale = sprite->getScale();

        data = std::make_tuple(scale.x, scale.y);

        return data;
    }
}

File: lua/ch2d/lib/sprite.lua
local Sprite = {}
Sprite.__index = Sprite

function Sprite.create()
	local sprite = {}
	setmetatable(sprite, Sprite)

	sprite.id = ch2d.sprite.create()

	return sprite
end

function Sprite:remove()
	return ch2d.sprite.remove(self.id)
end

function Sprite:draw()
	return ch2d.sprite.draw(self.id)
end

function Sprite:remove()
	return ch2d.sprite.remove(self.id)
end

function Sprite:setPosition(_x, _y)
	return ch2d.sprite.setPosition(self.id, _x, _y)
end

function Sprite:setOrigin(_x, _y)
	return ch2d.sprite.setOrigin(self.id, _x, _y)
end

function Sprite:setRotation(_degrees)
	return ch2d.sprite.setRotation(self.id, _degrees)
end

function Sprite:setScale(_x, _y)
	return ch2d.sprite.setScale(self.id, _x, _y)
end

function Sprite:setTexture(_textureId)
	return ch2d.sprite.setTexture(self.id, _textureId)
end

function Sprite:setTextureRect(_top, _left, _width, _height)
	return ch2d.sprite.setTextureRect(self.id, _top, _left, _width, _height)
end

function Sprite:getPosition()
	return ch2d.sprite.getPosition(self.id)
end

function Sprite:getOrigin()
	return ch2d.sprite.getOrigin(self.id)
end

function Sprite:getRotation()
	return ch2d.sprite.getRotation(self.id)
end

function Sprite:getScale()
	return ch2d.sprite.getScale(self.id)
end

return Sprite
